File: src/main/java/eu/timerertim/downlomatic/utils/WebScrapers.java
package eu.timerertim.downlomatic.utils;

import io.github.bonigarcia.wdm.WebDriverManager;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.firefox.FirefoxDriver;
import org.openqa.selenium.firefox.FirefoxDriverLogLevel;
import org.openqa.selenium.firefox.FirefoxOptions;

import java.util.logging.Level;

/**
 * Provides webscrapers for usage. Has to be initialized at least once in order to work.
 */
public class WebScrapers {
    private static JSoupDriver jSoupDriver;
    private static FirefoxDriver firefoxDriver;

    /**
     * Returns the WebDriver responsible for handling non js webscraping.
     *
     * @return the JSoupDriver
     */
    public static JSoupDriver noJavaScript() {
        return jSoupDriver;
    }

    /**
     * Returns the WebDriver responsible for handling JavaScript webscraping.
     *
     * @return the FirefoxDriver
     */
    public static WebDriver javaScript() {
        return firefoxDriver;
    }

    /**
     * If one of the WebDrivers happens to be closed, you can reinitialize them with this method.
     */
    public static void reInitialize() {
        if (firefoxDriver.toString().contains("null")) {
            initializeFirefoxDriver();
        }
        if (jSoupDriver.toString().contains("null")) {
            initializeJSoupDriver();
        }
    }

    /**
     * Closes both WebDrivers to release resources.
     */
    public static void close() {
        firefoxDriver.quit();
        jSoupDriver.quit();
    }

    /**
     * Initializes the WebScraper for handling (non-)JavaScript websites.
     */
    public static void initialize() {
        java.util.logging.Logger.getLogger("org.openqa.selenium").setLevel(Level.OFF);

        WebDriverManager.firefoxdriver().setup();

        initializeFirefoxDriver();
        initializeJSoupDriver();
    }

    private static void initializeFirefoxDriver() {
        FirefoxOptions options = new FirefoxOptions().setHeadless(true).setLogLevel(FirefoxDriverLogLevel.FATAL).addArguments("--log", "fatal");
        options.setCapability("marionette", true);
        (firefoxDriver = new FirefoxDriver(options)).setLogLevel(Level.OFF);
    }

    private static void initializeJSoupDriver() {
        jSoupDriver = new JSoupDriver();
    }
}

File: shared/temp/main/kotlin/eu/timerertim/downlomatic/utils/WebScrapers.kt
package eu.timerertim.downlomatic.utils

import io.github.bonigarcia.wdm.WebDriverManager
import org.openqa.selenium.WebDriver
import org.openqa.selenium.firefox.FirefoxDriver
import org.openqa.selenium.firefox.FirefoxDriverLogLevel
import org.openqa.selenium.firefox.FirefoxOptions
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Provides webscrapers for usage. Has to be initialized at least once in order to work.
 */
object WebScrapers {
    private var jSoupDriver: JSoupDriver? = null
    private var firefoxDriver: FirefoxDriver? = null

    /**
     * Returns the WebDriver responsible for handling non js webscraping.
     *
     * @return the JSoupDriver
     */
    @JvmStatic
    fun noJavaScript(): JSoupDriver? {
        if (jSoupDriver == null) {
            initializeJSoupDriver()
        }
        return jSoupDriver
    }

    /**
     * Returns the WebDriver responsible for handling JavaScript webscraping.
     *
     * @return the FirefoxDriver
     */
    @JvmStatic
    fun javaScript(): WebDriver? {
        if (firefoxDriver == null) {
            initializeFirefoxDriver()
        }
        return firefoxDriver
    }

    /**
     * If one of the WebDrivers happens to be closed, you can reinitialize them with this method.
     */
    fun reInitialize() {
        if (firefoxDriver.toString().contains("null")) {
            initializeFirefoxDriver()
        }
        if (jSoupDriver.toString().contains("null")) {
            initializeJSoupDriver()
        }
    }

    /**
     * Closes both WebDrivers to release resources.
     */
    @JvmStatic
    fun close() {
        firefoxDriver?.quit()
        jSoupDriver?.quit()
    }

    /**
     * Initializes the WebScraper for handling (non-)JavaScript websites.
     */
    @JvmStatic
    fun initialize() {
        initializeFirefoxDriver()
        initializeJSoupDriver()
    }

    private fun initializeFirefoxDriver() {
        WebDriverManager.firefoxdriver().setup()

        Logger.getLogger("org.openqa.selenium").level = Level.OFF
        val options =
            FirefoxOptions().setHeadless(true).setLogLevel(FirefoxDriverLogLevel.FATAL).addArguments("--log", "fatal")
        options.setCapability("marionette", true)
        FirefoxDriver(options).also { firefoxDriver = it }.setLogLevel(Level.OFF)
    }

    private fun initializeJSoupDriver() {
        jSoupDriver = JSoupDriver()
    }
}